contracts/extensions/collections/HasSlashIndicatorContract.sol

```diff
@@ -9,22 +9,39 @@ contract HasSlashIndicatorContract is IHasSlashIndicatorContract, HasProxyAdmin
   ISlashIndicator internal _slashIndicatorContract;
 
   modifier onlySlashIndicatorContract() {
-    if (slashIndicatorContract() != msg.sender) revert ErrCallerMustBeSlashIndicatorContract();
+    _requiresSlashIndicatorContract();
     _;
   }
 
+  function _requiresSlashIndicatorContract() internal view {
+    assembly {
+      if iszero(eq(caller(), sload(_slashIndicatorContract.slot))) {
+        mstore(0x00, 0xa2e7092c)
+        revert(0x1c, 0x04)
+      }
+    }
+  }
+
   /**
    * @inheritdoc IHasSlashIndicatorContract
    */
-  function slashIndicatorContract() public view override returns (address) {
-    return address(_slashIndicatorContract);
+  function slashIndicatorContract() public view override returns (address addr) {
+    assembly {
+      addr := sload(_slashIndicatorContract.slot)
+    }
   }
 
   /**
    * @inheritdoc IHasSlashIndicatorContract
    */
   function setSlashIndicatorContract(address _addr) external override onlyAdmin {
-    if (_addr.code.length == 0) revert ErrZeroCodeContract();
+    assembly {
+      if iszero(extcodesize(_addr)) {
+        /// @dev value is equal to bytes4(keccak256("ErrZeroCodeContract()"))
+        mstore(0x00, 0x7bcd5091)
+        revert(0x1c, 0x04)
+      }
```

Comment:
In general, I don't think optimizing emitting events by asm is a good idea. Since this will cause problems in later maintaining.